pkg/applicationserver/io/packages/alcsync/v1/package.go

```diff
@@ -86,22 +101,35 @@ func (a *alcsyncpkg) HandleUp(
 			continue
 		}
 		if err != nil {
-			logger.WithError(err).Debug("Failed to execute command")
+			logger.WithError(err).WithField("command_id", cmd.Code()).Error("Failed to execute command")
+			evt := makeExecutionFailedEvtBuilder(cmd.Code(), err)
+			eventBuilders = append(eventBuilders, evt)
 			continue
 		}
 		if result != nil {
 			results = append(results, result)
+			eventBuilders = append(eventBuilders, result.GetEvtSuccessfullyExecuted())
 		}
 	}
 	downlink, err := MakeDownlink(results, fPort)
 	if err != nil {
-		logger.WithError(err).Debug("Failed to create downlink from results")
-		return err
+		logger.WithError(err).Error("Failed to create downlink from results")
+		eventBuilders = append(eventBuilders, EvtALCSyncPkgFail.With(
+			events.WithIdentifiers(up.GetEndDeviceIds()),
+			events.WithData(err),
+		))
+		return a.publishAndExit(ctx, eventBuilders, err)
 	}
 	if err := a.server.DownlinkQueuePush(ctx, up.EndDeviceIds, []*ttnpb.ApplicationDownlink{downlink}); err != nil {
-		logger.WithError(err).Debug("Failed to push downlinks to queue")
+		logger.WithError(err).Error("Failed to push downlinks to queue")
 	}
-	return nil
+	return a.publishAndExit(ctx, eventBuilders, nil)
```

Comment:
You should use a `defer` instead. Otherwise you're not publishing on panics or other nasty cases.